framework/include/auxkernels/NodalPatchRecovery.h

```diff
@@ -0,0 +1,130 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef NODALPATCHRECOVERY_H
+#define NODALPATCHRECOVERY_H
+
+#include "AuxKernel.h"
+#include "FEProblem.h"
+
+/**
+ * This class uses patch recovery technique to recover material properties to smooth nodal fields
+ * Zienkiewicz, O. C. and Zhu, J. Z. (1992), The superconvergent patch recovery and a posteriori
+ * error estimates. Part 1: The recovery technique. Int. J. Numer. Meth. Engng., 33: 1331-1364.
+ * doi:10.1002/nme.1620330702
+ */
+class NodalPatchRecovery;
+
+template <>
+InputParameters validParams<NodalPatchRecovery>();
+
+class NodalPatchRecovery : public AuxKernel
+{
+public:
+  NodalPatchRecovery(const InputParameters & parameters);
+  virtual ~NodalPatchRecovery(){};
+  /// Override from Auxkernel to suppress warnings
+  template <typename T>
+  const MaterialProperty<T> & getMaterialProperty(const std::string & name);
+  /// Override from Auxkernel to suppress warnings
+  template <typename T>
+  const MaterialProperty<T> & getMaterialPropertyOld(const std::string & name);
+  /// Override from Auxkernel to suppress warnings
+  template <typename T>
+  const MaterialProperty<T> & getMaterialPropertyOlder(const std::string & name);
+
+protected:
+  /// generate all combinations of n choose k
+  virtual std::vector<std::vector<unsigned>> nchoosek(unsigned N, unsigned K);
+
+  /**
+   * generate a complete multi index table for given dimension and order
+   * i.e. given dim = 2, order = 2, generated table will have the following content
```

Comment:
Either here or in the documentation for _multi_index, explain what the entries mean.